File: 15-3sum/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int length = nums.size();
        vector<vector<int>> res;
        if (length < 3){
            return res;
        }
        sort(nums.begin(),nums.end());
        for(int i=0;i<length-2; ++i){
            if (i == 0 || nums[i] != nums[i-1]){
                int j = i+1,k=length -1;
                while(j<k){
                    int sum = nums[i] + nums[j] + nums[k];
                    if(sum == 0){
                        // vector<int> v1{nums[i],nums[2]}
                        res.push_back({nums[i],nums[j],nums[k]});
                        while(j<k && nums[j] == nums[j+1]){
                            j += 1;
                        }
                        while(j<k && nums[k] == nums[k-1]){
                            k -= 1;
                        }
                        j++;
                        k--;
                    } else if(sum < 0){
                        j += 1;
                    } else {
                        k -= 1;
                    }
                }
            }
        }
        return res;
    }
};

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 3: return []
        result = []
        nums.sort()
        for i in range(n-2):
            if i == 0 or nums[i] != nums[i-1]:
                j,k = i+1, n-1
                while j < k:
                    sum = nums[i] + nums[j] + nums[k]
                    if sum == 0:
                        result.append([nums[i], nums[j], nums[k]])
                        while j < k and nums[j] == nums[j+1]: j += 1
                        while j < k and nums[k] == nums[k-1]: k -= 1
                        j, k = j+1, k-1
                    elif sum < 0: j += 1
                    else: k -= 1
        return result